keps/sig-storage/1682-csi-driver-skip-permission/README.md

```diff
@@ -0,0 +1,238 @@
+`SupportsFSGroup` which can have following possible values:
+
+ - `OnlyRWO` --> Current behavior. Attempt to modify the volume ownership and permissions to the defined `fsGroup` when the volume is 
+ mounted if accessModes is RWO.
+ - `Never` --> New behavior. Attach the volume without attempting to modify volume ownership or permissions.
+ - `Always` --> New behavior. Always attempt to apply the defined fsGroup to modify volume ownership and permissions.
+
+```go
+type SupportsFsGroup string
+
+const(
+    OnlyRWO SupportsFsGroup = "OnlyRWO"
+    Always SupportsFsGroup = "Always"
+    Never SupportsFsGroup = "Never"
+)
+
+type CSIDriverSpec struct {
+    // SupportsFSGroup ← new field
+    // Defines if the underlying volume supports changing ownership and 
+    // permission of the volume before being mounted. 
+    // If set to Always, SupportsFSGroup indicates that 
+    // the volumes provisioned by this CSIDriver support volume ownership and 
+    // permission changes, and the filesystem will be modified to match the 
+    // defined fsGroup every time the volume is mounted.
+    // If set to Never, then the volume will be mounted without modifying
+    // the volume's ownership or permissions.
+    // Defaults to OnlyRWO, which results in the volume being examined
+    // and the volume ownership and permissions attempting to be updated
+    // only when the PodSecurityPolicy's fsGroup is explicitly defined and the
+    // PersistentVolumes's accessModes is RWO.
+    // + optional
+    SupportsFSGroup *SupportsFsGroup
+}
+```
+### Test Plan
+
+A test plan will include the following tests:
+
+* Basic tests including a permutation of the following values:
+  - CSIDriver.Spec.SupportsFSGroup  (`Always`/`Never`/`OnlyRWO`)
+  - PersistentVolumeClaim.Status.AccessModes (`ReadWriteOnly`, `ReadOnlyMany`,`ReadWriteMany`)
+* E2E tests
+
+### Graduation Criteria
+
+* Alpha in 1.19 provided all tests are passing.
+* All functionality is guarded by a new alpha `CSIVolumeSupportFSGroup` feature gate.
+
+* Beta in 1.20 with design validated by at least two customer deployments
+  (non-production), with discussions in SIG-Storage regarding success of
+  deployments. 
+* The `CSIVolumeSupportFSGroup` feature gate will graduate to beta.
+
+
+* GA in 1.21, with E2E tests in place tagged with feature Storage.
+* The `CSIVolumeSupportFSGroup` feature gate will graduate to GA.
+
+[issues]: https://github.com/kubernetes/enhancements/issues/1682
+
+## Production Readiness Review Questionnaire
+
+### Feature enablement and rollback
+* **How can this feature be enabled / disabled in a live cluster?**
+  - [x] Feature gate (also fill in values in `kep.yaml`)
+    - Feature gate name: CSIVolumeSupportFSGroup 
+    - Components depending on the feature gate: kubelet
+
+* **Does enabling the feature change any default behavior?**
+  Enabling the feature gate will **not** change the default behavior.
```

Comment:
My understanding is that if this feature is enabled, then the default is OnlyRWO (the volume will be examined and the volume ownership and permissions will be updated only when the PodSecurityPolicy's fsGroup is explicitly defined and the PersistentVolumes's accessModes is RWO).  

Without enabling this feature, volume permission and ownership change is done automatically for every volume (which is stated on line 89).

So it seems to me there is a change in default behavior?

Reply:
I am hesitant to call the default behaviour "OnlyRWO" but this is something we can figure out during implementation phase too. 

Even without this feature - the default behaviour is not that we automatically change permissions for every volume - https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/csi/csi_mounter.go#L379 . We only change permission of a CSI volume if pod has `fsGroup` and:
1. volume has `fsType` on it.
2. Volume has RWO access mode.

So using `OnlyRWO` elides fstype check.

Reply:
On line 121, it says "Defaults to OnlyRWO".  If that is not intended, please clarify what is the default value.

Please clarify the difference in default behavior in this section.

Reply:
I've updated the default behavior above. Using `OnlyRWO` should be identical to the current default behavior.

Reply:
One note - @gnufied is correct that we need `fsType` to be != "" as well. I've included this in both the current default behavior along with this section.